src/unix/lwt_unix_unix.h

```diff
@@ -1831,19 +1820,21 @@ static void worker_readdir_n(struct job_readdir_n *job)
 {
     long i;
     for (i = 0; i < job->count; i++) {
-        struct dirent *ptr;
-        int result = readdir_r(job->dir, &job->entries[i], &ptr);
+        errno = 0;
+        struct dirent *entry = readdir(job->dir);
 
         /* An error happened. */
-        if (result != 0) {
-            job->error_code = result;
+        if (entry == NULL && errno != 0) {
+            job->error_code = errno;
             return;
         }
 
         /* End of directory reached */
-        if (ptr == NULL) break;
-    }
+        if (entry == NULL && errno == 0) break;
 
+        /* All is good */
+        job->entries[i] = *entry;
```

Comment:
I almost want to use`caml_copy_string` there. It sort of breaks the pattern of “`worker` does the C/system part and `result` does the conversion.” And on top of that, I'm not sure what happens if you call `caml_copy_string` and then abandon those values because of an error. Do you need to free them?

Actually… see other comment.

I see little point in copying the string twice: once in C and then once to convert to OCaml layout. Any opinions about it? Any comments on the paragraph above?